File: cs214/projects/proj07/clojure/src/names.clj
;;;; names.clj creates and tests a clojure Name type.
;;;;
;;;; Output: results of testing Name functions.
;;;;
;;;; Usage: clojure -m names
;;;;
;;;; Begun by: Dr. Adams, CS 214 at Calvin College.
;;;; Completed by: Advait Scaria
;;;; Date: 04/02/2019
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(ns names)   ; program name

;;; define a Clojure record named Name 
;;; (which compiles to a Java class).

(defrecord Name [firstName middleName lastName])


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; make-Name constructs a Name from three strings. 
;;; Receive: firstN, middleN and lastN, three strings.
;;; Return: the Name (firstN middleN lastN). 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn make-Name [first middle last]
	(->Name first middle last)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; getFirst() extracts the first name of a Name object.
;;; Receive: aName, a Name.
;;; Return: the firstName string in aName.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(defn getFirst [^Name aName]
	(:firstName aName)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; project 7 method
;;; setFirst() sets the first name of a Name object.
;;; Receive: aName, a Name & newF, a String.
;;; POST: set firstName equal to newF.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn setFirst [^Name aName newF]
  (->Name newF (:middleName aName) (:lastName aName))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; getMiddle() extracts the middle name of a name object.
;;; Receive: aName, a Name. 
;;; Return: the middleName string in aName.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
(defn getMiddle [^Name aName]
	(:middleName aName)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; project 7 method
;;; setMiddle() sets the middle name of a Name object.
;;; Receive: aName, a Name & newM, a String.
;;; POST: set middleName equal to newM.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn setMiddle [^Name aName newM]
  (->Name (:firstName aName) newM (:lastName aName))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; getLast() extracts the last name of a name object.
;;; Receive: aName, a Name. 
;;; Return: the lastName string in aName.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
(defn getLast [^Name aName]
	(:lastName aName)
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; project 7 method
;;; setLast() sets the last name of a Name object.
;;; Receive: aName, a Name & newL, a String.
;;; POST: set lastName equal to newL.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn setLast [^Name aName newL]
  (->Name (:firstName aName) (:middleName aName) newL)
)


;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; toString() returns a full name in F-M-L order. 
;;; Receive: aName, a Name. 
;;; Return: firstName, middleName, lastName,
;;;           separated by spaces. 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
(defn toString [^Name aName]
	(str (getFirst aName) " " (getMiddle aName) " " (getLast aName))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; printName() displays a name object.
;;; Receive: aName, a Name.
;;; Output: the strings in aName.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
(defn printName [^Name aName]
	(print (toString aName))
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; project 7 method
;;; toString() returns a full name in F-M-L order. 
;;; Receive: aName, a Name. 
;;; Return: firstName, middleName, lastName,
;;;           separated by spaces. 
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
(defn lfmi [^Name aName]
	(str (getLast aName) ", " (getFirst aName) " " (get (getMiddle aName) 0) ".")
)

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; project 7 method
;;; read() reads name inputs and populates name object with valid inputs. 
;;; Receive: aName, a Name. 
;;; POST: Valid name inputs read from user input and
;;;       Name object filled with input values for
;;;       first, middle, and last names.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 
(defn readInput []
		(do (print "Enter the full name: ") (flush)
			(let 
				[
					myFirst (str (read))
					myMiddle (str (read))
					myLast (str (read))
				]
				(->Name myFirst myMiddle myLast)
			)
		)
) 

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; A simple driver to test our Name functions. 
;;; Output: the result of testing our Name functions.
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn -main []
  (let
    [                                          ; 3 ways to construct an object:
      name1 (make-Name "John" "Paul" "Jones")  ; -using our "make-" constructor
      name2 (->Name "Jane" "Penelope" "Jones") ; -invoking constructor directly
                                               ; -mapping field-names to values
      name3 (map->Name {:lastName "Jones" :firstName "Jinx" :middleName "Joy"})

      name4 (setFirst name1 "Jake")												; 3 new names objects set for testing setters
    		name5 (setMiddle name4 "Joseph")
    		name6 (setLast name5 "Paul")

    		name7 (readInput)
    ]
    ;; ----- SECTION 1 -----
    (println)
    (print name1) (println)
    (assert (= (getFirst name1) "John") "getFirst(1) failed")
    (assert (= (getMiddle name1) "Paul") "getMiddle(1) failed")
    (assert (= (getLast name1) "Jones") "getLast(1) failed")
    (assert (= (toString name1) "John Paul Jones") "toString(1) failed")
    (printName name1) (println)
    ;; ----- SECTION 2 -----
    (println)
    (print name2) (println)
    (assert (= (getFirst name2) "Jane") "getFirst(2) failed")
    (assert (= (getMiddle name2) "Penelope") "getMiddle(2) failed")
    (assert (= (getLast name2) "Jones") "getLast(2) failed")
    (assert (= (toString name2) "Jane Penelope Jones") "toString(2) failed")
    (printName name2) (println)
    ;; ----- SECTION 3 -----
    (println)
    (print name3) (println)
    (assert (= (getFirst name3) "Jinx") "getFirst(3) failed")
    (assert (= (getMiddle name3) "Joy") "getMiddle(3) failed")
    (assert (= (getLast name3) "Jones") "getLast(3) failed")
    (assert (= (toString name3) "Jinx Joy Jones") "toString(3) failed")
    (printName name3) (println)

    ;; ----- PROJECT 7 METHODS' TESTS -----

    ;; setFirst(), setMiddle(), setLast() tests
    (assert (= (getFirst name4) "Jake") "setFirst() failed")
    (assert (= (getMiddle name5) "Joseph") "setMiddle() failed")
    (assert (= (getLast name6) "Paul") "setLast() failed")
    (println) (print "PROJECT 7 METHODS' TESTS BELOW: ") (println)
    (printName name4) (println)
    (printName name5) (println)
    (printName name6) (println)

    ;; lfmi() test
    (assert (= (lfmi name6) "Paul, Jake J.") "lfmi() failed")

    ;; read() test -- the name Dawson Alexander Buist must be input when initializing test7 through read() in let[]
    (println)
    (assert (= (getFirst name7) "Dawson") "read() failed")
    (assert (= (getMiddle name7) "Alexander") "read() failed")
    (assert (= (getLast name7) "Buist") "read() failed")
    (printName name7) (println)

    (println "\nAll tests passed!\n")
  )
)

File: cs214/projects/proj07/Names.java
/* Names.java is a 'driver' to test class Name.
 * Begun by: Prof. Adams, for CS 214 at Calvin College.
 * Student: Advait Scaria
 * Date: 03/29/2019
 ************************************************************/

import java.util.Scanner;

public class Names  
{
  public static void main(String[] args)
  {
      Name aName = new Name("John", "Paul", "Jones");

	    assert aName.getFirst().equals("John");
      assert aName.getMiddle().equals("Paul");
      assert aName.getLast().equals("Jones");

      System.out.println(aName);
      assert aName.toString().equals("John Paul Jones");

      // ------ PROJECT 7 METHODS' TESTS ------
      
      //setFirst() test
      aName.setFirst("Jake");
      assert aName.getFirst().equals("Jake");

      //setMiddle() test
      aName.setMiddle("Joseph");
      assert aName.getMiddle().equals("Joseph");

      //setLast() test
      aName.setLast("Paul");
      assert aName.getLast().equals("Paul");

      //lfmi() test
      assert aName.lfmi().equals("Paul, Jake J.");

      //read() test
      aName.read();
      assert aName.getFirst().equals("Dawson");     //default name that will be input in console to test
      assert aName.getMiddle().equals("Alexander");
      assert aName.getLast().equals("Buist");
      System.out.println(aName);    //visually confirm read works 

      System.out.println("All tests passed!");
  }
}

class Name
   {
      private String myFirst,
                     myMiddle,
                     myLast;

      /* Name explicit constructor...constructs a Name object from three Strings.
       * Receive: first, middle and last, three Strings.
       * POST: Makes a Name object and initializes its three variables to the given parameters.
       */
      public Name(String first, String middle, String last)
   		{
      		myFirst = first;
      		myMiddle = middle;
      		myLast = last;
   		}

   	  /* getFirst()
   	   * PRE: Name is initialized with its three name parts.
       * Return: myFirst, the first name of the Name object.
       */
   	  public String getFirst()
   		{
      		return myFirst;
   		}

      /* setFirst()...project 7 method
       * Receive: first, a String
       * PRE: Valid String passed as first name.
       * POST: set myFirst equal to first.
       */
      public void setFirst(String first)
      {
          myFirst = first;
      }

      /* getMiddle()
   	   * PRE: Name is initialized with its three name parts.
       * Return: myMiddle, the middle name of the Name object.
       */
   	  public String getMiddle()
   		{
      		return myMiddle;
   		}

      /* setMiddle()...project 7 method
       * Receive: middle, a String
       * PRE: Valid String passed as middle name.
       * POST: set myMiddle equal to middle.
       */
      public void setMiddle(String middle)
      {
          myMiddle = middle;
      }

   	  /* getLast()
   	   * PRE: Name is initialized with its three name parts.
       * Return: myLast, the last name of the Name object.
       */
   	  public String getLast()
   		{
      		return myLast;
   		}

      /* setLast()...project 7 method
       * Receive: last, a String
       * PRE: Valid String passed as last name.
       * POST: set myLasr equal to last.
       */
      public void setLast(String last)
      {
          myLast = last;
      }

   	  /* toString()
   	   * PRE: Name is initialized with its three name parts.
       * Return: a String representation of the Name object
       *		 that containts all of its three names in order 
       *     separated by spaces.
       */
   	  public String toString()
   		{
      		return myFirst + ' ' + myMiddle + ' ' + myLast;
   		}

      /* lfmi()...project 7 method
       * PRE: Name is initialized with its three name parts.
       * Return: a String representation of the Name object
       *     that containts all of its three names in the 
       *     order of last name, first name, middle initial.
       */
      public String lfmi()
      {
          return myLast + ", " + myFirst + " " + myMiddle.charAt(0) + ".";  
      }

      /* read()...project 7 method
       * POST: Valid name inputs read from user input and
       *       Name object filled with input values for
       *       first, middle, and last names.
       */
      public void read() 
      {
          Scanner sc = new Scanner(System.in);
          String first = sc.nextLine();
          setFirst(first);
          String middle = sc.nextLine();
          setMiddle(middle);
          String last = sc.nextLine();
          setLast(last);
      }
   };
